File: viewer/src/geometry.ts
export type Rect = Readonly<{ x: number; y: number; width: number; height: number }>;
export type Size = Readonly<{ width: number; height: number }>;
export type Point = Readonly<{ x: number; y: number }>;

export function scaleRect(rect: Rect, source: Size, destination: Size): Rect {
  const scaleX = destination.width / source.width;
  const scaleY = destination.height / source.height;
  return {
    x: rect.x * scaleX,
    y: rect.y * scaleY,
    width: rect.width * scaleX,
    height: rect.height * scaleY,
  };
}

export function containsPoint(rect: Rect, point: Point): boolean {
  return point.x >= rect.x
    && point.y >= rect.y
    && point.x <= rect.x + rect.width
    && point.y <= rect.y + rect.height;
}

File: generator/src/Factory/Geometry.hs
-- | Pure coordinate and matrix operations.
--
-- PDF places (0,0) at the bottom-left. Browsers place it at the top-left.
-- Keeping this conversion in one module gives the project one authoritative
-- answer to the most common source of misplaced links and images.
module Factory.Geometry
  ( applyMatrix
  , boardMatrix
  , identityMatrix
  , multiplyMatrix
  , pdfPointToBoard
  , rectangleToBoard
  ) where

import Factory.Domain
  ( BoardSpace
  , Coordinate (Coordinate, unCoordinate)
  , Matrix (Matrix)
  , PdfSpace
  , Point (Point)
  , Rect (Rect)
  )

identityMatrix :: Matrix
identityMatrix = Matrix 1 0 0 1 0 0

-- | Apply the left matrix first, matching the PDF @cm@ operator.
multiplyMatrix :: Matrix -> Matrix -> Matrix
multiplyMatrix (Matrix a1 b1 c1 d1 e1 f1) (Matrix a2 b2 c2 d2 e2 f2) =
  Matrix
    (a1 * a2 + b1 * c2)
    (a1 * b2 + b1 * d2)
    (c1 * a2 + d1 * c2)
    (c1 * b2 + d1 * d2)
    (e1 * a2 + f1 * c2 + e2)
    (e1 * b2 + f1 * d2 + f2)

applyMatrix :: Matrix -> Point space -> Point space
applyMatrix (Matrix a b c d e f) (Point (Coordinate x) (Coordinate y)) =
  Point (Coordinate (a * x + c * y + e)) (Coordinate (b * x + d * y + f))

pdfPointToBoard :: Coordinate PdfSpace -> Point PdfSpace -> Point BoardSpace
pdfPointToBoard pageHeight (Point x y) =
  Point (coerceCoordinate x) (coerceCoordinate pageHeight - coerceCoordinate y)

-- | Convert a PDF matrix into the equivalent CSS matrix.
boardMatrix :: Coordinate PdfSpace -> Matrix -> Matrix
boardMatrix (Coordinate pageHeight) (Matrix a b c d e f) =
  Matrix a (-b) c (-d) e (pageHeight - f)

rectangleToBoard :: Coordinate PdfSpace -> Rect PdfSpace -> Rect BoardSpace
rectangleToBoard height (Rect x y width rectangleHeight) =
  Rect
    (coerceCoordinate x)
    (coerceCoordinate height - coerceCoordinate y - coerceCoordinate rectangleHeight)
    (coerceCoordinate width)
    (coerceCoordinate rectangleHeight)

coerceCoordinate :: Coordinate from -> Coordinate to
coerceCoordinate = Coordinate . unCoordinate
